code/drasil-lang/lib/Language/Drasil/Expr/Lang.hs

```diff
@@ -317,178 +333,183 @@ instance Typed Expr Space where
-    (_               , Right e) -> Right e
-    (Right e         , _      ) -> Right e
-  infer cxt (LABinaryOp IndexOf l n) = case (infer cxt l, infer cxt n) of
-    (Left (S.Set lt), Left nt) -> if S.isBasicNumSpace lt && nt == lt-- I guess we should only want it to be natural numbers, but integers or naturals is fine for now
-      then Left lt
-      else Right $ "List accessor not of type Integer nor Natural, but of type `" ++ show nt ++ "`"
-    (Left lt         , Left _)  -> Right $ "List accessor expects a list/vector, but received `" ++ show lt ++ "`."
-    (_               , Right e) -> Right e
-    (Right e         , _      ) -> Right e
-  infer cxt (OrdBinaryOp _ l r) = case (infer cxt l, infer cxt r) of
-    (Left lt, Left rt) -> if S.isBasicNumSpace lt && lt == rt
-      then Left S.Boolean
-      else Right $ "Both operands of a numeric comparison must be the same numeric type, got: " ++ show lt ++ ", " ++ show rt ++ "."
-    (_, Right re) -> Right re
-    (Right le, _) -> Right le
-
-  infer cxt (NVVBinaryOp Scale l r) = case (infer cxt l, infer cxt r) of
-    (Left lt, Left (S.Vect rsp)) -> if S.isBasicNumSpace lt && lt == rsp
-      then Left rsp
-      else if lt /= rsp then
-        Right $ "Vector scaling expects a scaling by the same kind as the vector's but found scaling by`" ++ show lt ++ "` over vectors of type `" ++ show rsp ++ "`."
-      else
-        Right $ "Vector scaling expects a numeric scaling, but found `" ++ show lt ++ "`."
-    (Left _, Left rsp) -> Right $ "Vector scaling expects vector as second operand. Received `" ++ show rsp ++ "`."
-    (_, Right rx) -> Right rx
-    (Right lx, _) -> Right lx
+
+  infer cxt (ArithBinaryOp Subt l r) = do
+    lt <- infer cxt l
+    rt <- infer cxt r
+    assertEquivNumeric
+      lt rt
+      (\ls rs -> "Subtraction should only be applied to the same numeric typed operands. Received `" ++ ls ++ "` - `" ++ rs ++ "`.")
+    pure lt
+
+  infer cxt (BoolBinaryOp _ l r) = do
+    lt <- infer cxt l
+    rt <- infer cxt r
+    let msg = const $ "Boolean expression contains non-boolean operand. Received `" ++ show lt ++ "` & `" ++ show rt ++ "`."
+    assertBoolean lt msg
+    assertBoolean rt msg
+    pure S.Boolean
+
+  infer cxt (EqBinaryOp _ l r) = do
+    lt <- infer cxt l
+    rt <- infer cxt r
+    lt ~== rt $ \lsp rsp -> "Both operands of an (in)equality (=/≠) must be of the same type. Received `" ++ lsp ++ "` & `" ++ rsp ++ "`."
+    pure S.Boolean
+
+  infer cxt (LABinaryOp Index l n) = do
+    lt <- infer cxt l
+    vet <- assertVector lt (\sp -> "List accessor expects a vector, but received `" ++ sp ++ "`.")
+
+    rt <- infer cxt n
+    assertIndexLike rt
+      (\sp -> "List accessor expects an index-like type (Integer or Natural), but received `" ++ sp ++ "`.")
+
+    pure vet
+
+  -- TODO: Why do we have *both* Index and IndexOf?
+  infer cxt (LABinaryOp IndexOf l n) = infer cxt (LABinaryOp Index l n)
```

Comment:
A thing to follow-up on later.

Reply:
Thanks! Fixed in 857a77af6461e85098be21d6ee57695855cd4012 and 67f235e9ff81b9ddd4c51ecfa05aad3e91f079a9 now.